src/lib/bundleUtils.ts

```diff
@@ -0,0 +1,213 @@
+import { Interface, Transaction, TransactionRequest, Wallet } from "ethers";
+import express from "express";
+import { FlashbotsBundleProvider } from "flashbots-ethers-v6-provider-bundle";
+import { getBaseFee, getMaxBlockByChainId, getProvider, initWallets } from "./helpers";
+
+import MevShareClient, { BundleParams } from "@flashbots/mev-share-client";
+import { JSONRPCID, createJSONRPCSuccessResponse } from "json-rpc-2.0";
+
+import { ovalAbi } from "../abi";
+import { env } from "./env";
+import { Logger } from "./logging";
+import { Refund } from "./types";
+const { ovalConfigs } = env;
+
+export const ovalInterface = Interface.from(ovalAbi);
+
+export const createUnlockLatestValueTx = async (
+  wallet: Wallet,
+  baseFee: bigint,
+  data: string,
+  chainId: bigint,
+  ovalAddress: string,
+) => {
+  const nonce = await wallet.getNonce();
+
+  // Construct transaction to call unlockLatestValue on Oval Oracle from permissioned address.
+  const unlockTx: TransactionRequest = {
+    type: 2,
+    chainId,
+    to: ovalAddress,
+    nonce,
+    value: 0,
+    gasLimit: 200000,
+    data,
+    // Double the current base fee as a basic safe gas estimate. We can make this more sophisticated in the future.
+    maxFeePerGas: baseFee * 2n,
+    maxPriorityFeePerGas: 0, // searcher should pay the full tip.
+  };
+
+  const signedUnlockTx = await wallet.signTransaction(unlockTx);
+
+  const unlockTxHash = Transaction.from(signedUnlockTx).hash;
+  if (!unlockTxHash) throw new Error("No hash in signed unlock transaction");
+
+  return { unlockTxHash, signedUnlockTx };
+};
+
+// Prepare unlockLatestValue transaction for a given Oval instance and simulate the bundle with the unlock transaction prepended if simulate is true.
+export const prepareUnlockTransaction = async (
+  flashbotsBundleProvider: FlashbotsBundleProvider,
+  backrunTxs: string[],
+  targetBlock: number,
+  ovalAddress: string,
+  req: express.Request,
+  simulate = true,
+) => {
+  const provider = getProvider();
+  const unlockerWallets = initWallets(provider);
```

Comment:
these are the same for every requests. can we better set them up when starting node

Reply:
though, maybe not a big deal after all as they are sync